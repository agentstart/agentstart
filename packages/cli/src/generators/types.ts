/* agent-frontmatter:start
AGENT: CLI generator types
PURPOSE: Defines shared generator type signatures for CLI schema emitters.
USAGE: Import to enforce consistent generator contracts.
EXPORTS: SchemaGenerator
FEATURES:
  - Specifies output payload shape for schema generators
  - Ensures adapters provide consistent generator options
SEARCHABLE: packages, cli, src, generators, types, generator
agent-frontmatter:end */

import type { MemoryAdapter } from "@agentstart/types";
import type { AgentStartOptions } from "agentstart";

export type SchemaGenerator = (opts: {
  file?: string;
  adapter: MemoryAdapter;
  options: Omit<AgentStartOptions, "agent">;
}) => Promise<{
  code?: string;
  fileName: string;
  overwrite?: boolean;
  append?: boolean;
}>;
