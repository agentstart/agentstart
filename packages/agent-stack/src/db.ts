/* agent-frontmatter:start
AGENT: Agent Stack database exports
PURPOSE: Surface database helpers from the infra package through the main bundle
USAGE: import { getMigrationPlan } from "agent-stack/db"
EXPORTS: database helpers, schema utilities
FEATURES:
  - Re-exports infra database modules for consumers
  - Keeps DB layer colocated with the primary package entrypoints
SEARCHABLE: database exports, infra db, agent stack db
agent-frontmatter:end */

export * from "@agent-stack/infra/db";
