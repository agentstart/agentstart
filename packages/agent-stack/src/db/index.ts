/* agent-frontmatter:start
AGENT: Database helper exports
PURPOSE: Re-export shared database helpers for adapter implementations
USAGE: import { getTables } from "agent-stack/db"
EXPORTS: field helpers, getTables, toZodSchema, getMigrations, getSchema, convert utilities, schema types
FEATURES:
  - Centralizes database helper exports for adapters and tests
  - Excludes plugin-specific helpers to keep API focused
SEARCHABLE: database helpers, adapter db exports, getTables
agent-frontmatter:end */

export * from "./adapter";
export * from "./field";
export * from "./get-migration";
export * from "./get-schema";
export * from "./get-tables";
export * from "./schema";
export * from "./to-zod";
export * from "./utils";
