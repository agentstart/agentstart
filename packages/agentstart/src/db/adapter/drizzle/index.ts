/* agent-frontmatter:start
AGENT: Agent runtime tool module
PURPOSE: Implements Packages tool execution within the AgentStart runtime.
USAGE: Register the "packages" tool when composing the agent configuration to expose this capability.
EXPORTS: Re-exports module symbols
FEATURES:
  - Bridges sandbox APIs into the Packages workflow
  - Streams structured progress updates and normalizes tool output
SEARCHABLE: packages, agentstart, src, db, adapter, drizzle, index, tool, agent, runtime
agent-frontmatter:end */

export * from "./drizzle-adapter";
