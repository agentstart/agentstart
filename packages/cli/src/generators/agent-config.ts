/* agent-frontmatter:start
AGENT: Agent Start CLI generator
PURPOSE: Compose Agent Start configuration files with optional database wiring
USAGE: await generateConfig({ current_user_config, format, spinner, database })
EXPORTS: generateConfig, Import
FEATURES:
  - Formats existing agent config using the caller supplied formatter
  - Injects adapter imports and database configuration when requested
  - Provides string helpers for safe insertion without full AST parsing
SEARCHABLE: cli generator, agent config, database generator
agent-frontmatter:end */

import { logger } from "@agentstart/utils";
import type { spinner as clackSpinner } from "@clack/prompts";
import type { SupportedDatabases } from "../commands/init";

export type Import = {
  path: string;
  variables:
    | { asType?: boolean; name: string; as?: string }[]
    | { asType?: boolean; name: string; as?: string };
};

type Format = (code: string) => Promise<string>;

type CommonIndexConfig = {
  regex: RegExp;
  getIndex: (args: {
    matchIndex: number;
    match: RegExpMatchArray;
  }) => number | null;
};

const START_OF_AGENT_START: CommonIndexConfig = {
  regex: /defineAgentConfig\({()/m,
  getIndex: ({ matchIndex }) => matchIndex + "defineAgentConfig({".length,
};

const getErrorMessage = (error: unknown) => {
  if (error instanceof Error) return error.message;
  if (typeof error === "string") return error;
  try {
    return JSON.stringify(error);
  } catch {
    return "Unknown error";
  }
};

export async function generateConfig({
  format,
  current_user_config,
  spinner,
  database,
}: {
  format: Format;
  current_user_config: string;
  spinner: ReturnType<typeof clackSpinner>;
  database: SupportedDatabases | null;
}): Promise<{
  generatedCode: string;
  dependencies: string[];
  envs: string[];
}> {
  // Central place for side-effectful mutations applied to the user config string.
  const configGeneration = {
    // Add import statements before formatting so downstream generators stay readable.
    addImport: async (opts: {
      imports: Import[];
      config: string;
    }): Promise<{ code: string; dependencies: string[]; envs: string[] }> => {
      let importString = "";
      for (const import_ of opts.imports) {
        if (Array.isArray(import_.variables)) {
          importString += `import { ${import_.variables
            .map(
              (x) =>
                `${x.asType ? "type " : ""}${x.name}${
                  x.as ? ` as ${x.as}` : ""
                }`,
            )
            .join(", ")} } from "${import_.path}";\n`;
        } else {
          importString += `import ${import_.variables.asType ? "type " : ""}${
            import_.variables.name
          }${import_.variables.as ? ` as ${import_.variables.as}` : ""} from "${
            import_.path
          }";\n`;
        }
      }
      try {
        const newContent = await format(importString + opts.config);
        return { code: newContent, dependencies: [], envs: [] };
      } catch (error) {
        throw new Error(
          `Failed to generate new agent config during import addition phase.`,
          { cause: error },
        );
      }
    },
    // Inject the database adapter snippet and capture follow-up dependencies/envs.
    addDatabase: async (opts: {
      database: SupportedDatabases;
      config: string;
    }): Promise<{ code: string; dependencies: string[]; envs: string[] }> => {
      const requiredEnvs: string[] = [];
      const requiredDeps: string[] = [];
      let databaseCode = "";

      const addDb = async ({
        db_code,
        dependencies,
        envs,
        imports,
        code_before_agentStart,
      }: {
        imports: Import[];
        db_code: string;
        envs: string[];
        dependencies: string[];
        code_before_agentStart?: string;
      }) => {
        if (code_before_agentStart) {
          const startOfAgentStart = getGroupInfo(
            opts.config,
            START_OF_AGENT_START,
          );
          if (!startOfAgentStart) {
            throw new Error(
              "[addDb] Couldn't find start of agentStart() function.",
            );
          }
          opts.config = insertContent({
            line: startOfAgentStart.line - 1,
            character: 0,
            content: opts.config,
            insert_content: `\n${code_before_agentStart}\n`,
          });
        }

        const codeGen = await configGeneration.addImport({
          config: opts.config,
          imports,
        });
        opts.config = codeGen.code;
        databaseCode = db_code;
        requiredEnvs.push(...envs, ...codeGen.envs);
        requiredDeps.push(...dependencies, ...codeGen.dependencies);
      };

      if (opts.database === "sqlite") {
        await addDb({
          db_code: `new Database(process.env.DATABASE_URL || "database.sqlite")`,
          dependencies: ["better-sqlite3"],
          envs: ["DATABASE_URL"],
          imports: [
            {
              path: "better-sqlite3",
              variables: {
                asType: false,
                name: "Database",
              },
            },
          ],
        });
      } else if (opts.database === "postgres") {
        await addDb({
          db_code: `new Pool({\nconnectionString: process.env.DATABASE_URL || "postgresql://postgres:password@localhost:5432/database"\n})`,
          dependencies: ["pg"],
          envs: ["DATABASE_URL"],
          imports: [
            {
              path: "pg",
              variables: [
                {
                  asType: false,
                  name: "Pool",
                },
              ],
            },
          ],
        });
      } else if (opts.database === "mysql") {
        await addDb({
          db_code: `createPool(process.env.DATABASE_URL!)`,
          dependencies: ["mysql2"],
          envs: ["DATABASE_URL"],
          imports: [
            {
              path: "mysql2/promise",
              variables: [
                {
                  asType: false,
                  name: "createPool",
                },
              ],
            },
          ],
        });
      } else if (opts.database === "mssql") {
        const dialectCode = `new MssqlDialect({
            tarn: {
              ...Tarn,
              options: {
              min: 0,
              max: 10,
              },
            },
            tedious: {
              ...Tedious,
              connectionFactory: () => new Tedious.Connection({
              authentication: {
                options: {
                password: 'password',
                userName: 'username',
                },
                type: 'default',
              },
              options: {
                database: 'some_db',
                port: 1433,
                trustServerCertificate: true,
              },
              server: 'localhost',
              }),
            },
          })`;
        await addDb({
          code_before_agentStart: dialectCode,
          db_code: `dialect`,
          dependencies: ["tedious", "tarn", "kysely"],
          envs: ["DATABASE_URL"],
          imports: [
            {
              path: "tedious",
              variables: {
                name: "*",
                as: "Tedious",
              },
            },
            {
              path: "tarn",
              variables: {
                name: "*",
                as: "Tarn",
              },
            },
            {
              path: "kysely",
              variables: [
                {
                  name: "MssqlDialect",
                },
              ],
            },
          ],
        });
      } else if (
        opts.database === "drizzle:mysql" ||
        opts.database === "drizzle:sqlite" ||
        opts.database === "drizzle:pg"
      ) {
        await addDb({
          db_code: `drizzleAdapter(db, {\nprovider: "${opts.database.replace(
            "drizzle:",
            "",
          )}",\n})`,
          dependencies: [],
          envs: [],
          imports: [
            {
              path: "agentstart/db",
              variables: [
                {
                  name: "drizzleAdapter",
                },
              ],
            },
            {
              path: "./database.ts",
              variables: [
                {
                  name: "db",
                },
              ],
            },
          ],
        });
      } else if (
        opts.database === "prisma:mysql" ||
        opts.database === "prisma:sqlite" ||
        opts.database === "prisma:postgresql"
      ) {
        await addDb({
          db_code: `prismaAdapter(client, {\nprovider: "${opts.database.replace(
            "prisma:",
            "",
          )}",\n})`,
          dependencies: [`@prisma/client`],
          envs: [],
          code_before_agentStart: "const client = new PrismaClient();",
          imports: [
            {
              path: "agentstart/db",
              variables: [
                {
                  name: "prismaAdapter",
                },
              ],
            },
            {
              path: "@prisma/client",
              variables: [
                {
                  name: "PrismaClient",
                },
              ],
            },
          ],
        });
      } else if (opts.database === "mongodb") {
        await addDb({
          db_code: `mongodbAdapter(db)`,
          dependencies: ["mongodb"],
          envs: [`DATABASE_URL`],
          code_before_agentStart: [
            `const client = new MongoClient(process.env.DATABASE_URL || "mongodb://localhost:27017/database");`,
            `const db = client.db();`,
          ].join("\n"),
          imports: [
            {
              path: "agentstart/db",
              variables: [
                {
                  name: "mongodbAdapter",
                },
              ],
            },
            {
              path: "mongodb",
              variables: [
                {
                  name: "MongoClient",
                },
              ],
            },
          ],
        });
      }

      const startOfAgentStart = getGroupInfo(opts.config, START_OF_AGENT_START);
      if (!startOfAgentStart) {
        throw new Error(
          "[addDatabase] Couldn't find start of agentStart() function.",
        );
      }
      const newContent = insertContent({
        line: startOfAgentStart.line,
        character: startOfAgentStart.character,
        content: opts.config,
        insert_content: `memory: ${databaseCode},`,
      });

      try {
        const formatted = await format(newContent);
        return {
          code: formatted,
          dependencies: requiredDeps,
          envs: requiredEnvs,
        };
      } catch (error) {
        throw new Error(
          `Failed to generate new agent config during database addition phase.`,
          { cause: error },
        );
      }
    },
  };

  let newUserConfig: string;
  try {
    newUserConfig = await format(current_user_config);
  } catch (error) {
    spinner.stop(
      `Failed to format agent config before database generation.`,
      1,
    );
    logger.error(getErrorMessage(error));
    process.exit(1);
  }

  const totalDependencies: string[] = [];
  const totalEnvs: string[] = [];

  if (database) {
    try {
      const { code, dependencies, envs } = await configGeneration.addDatabase({
        config: newUserConfig,
        database,
      });
      newUserConfig = code;
      totalDependencies.push(...dependencies);
      totalEnvs.push(...envs);
    } catch (error) {
      spinner.stop(
        `Something went wrong while generating/updating your new agent config file.`,
        1,
      );
      logger.error(getErrorMessage(error));
      process.exit(1);
    }
  }

  return {
    generatedCode: newUserConfig,
    dependencies: totalDependencies,
    envs: totalEnvs,
  };
}

function insertContent(params: {
  line: number;
  character: number;
  content: string;
  insert_content: string;
}): string {
  const { line, character, content, insert_content } = params;
  const lines = content.split("\n");

  if (line < 1 || line > lines.length) {
    throw new Error("Invalid line number");
  }

  const targetLineIndex = line - 1;

  if (character < 0 || character > lines[targetLineIndex]!.length) {
    throw new Error("Invalid character index");
  }

  const targetLine = lines[targetLineIndex]!;
  const updatedLine =
    targetLine.slice(0, character) +
    insert_content +
    targetLine.slice(character);
  lines[targetLineIndex] = updatedLine;

  return lines.join("\n");
}

function getGroupInfo(
  content: string,
  commonIndexConfig: CommonIndexConfig,
): {
  line: number;
  character: number;
  index: number;
} | null {
  const { regex, getIndex } = commonIndexConfig;
  const match = regex.exec(content);
  if (!match) return null;

  const groupIndex = getIndex({ matchIndex: match.index, match });
  if (groupIndex === null) return null;

  const position = getPosition(content, groupIndex);
  return {
    line: position.line,
    character: position.character,
    index: groupIndex,
  };
}

const getPosition = (str: string, index: number) => {
  const lines = str.slice(0, index).split("\n");
  return {
    line: lines.length,
    character: lines[lines.length - 1]!.length,
  };
};
