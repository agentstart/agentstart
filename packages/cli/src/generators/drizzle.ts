/* agent-frontmatter:start
AGENT: Agent Stack CLI drizzle generator
PURPOSE: Produce Drizzle ORM schema files from Agent Stack metadata
USAGE: await generateDrizzleSchema({ adapter, options, file })
EXPORTS: generateDrizzleSchema
FEATURES:
  - Maps Agent Stack models into drizzle table definitions
  - Respects adapter options like provider and pluralization
  - Generates or updates agent schema files on demand
SEARCHABLE: drizzle schema, cli generator, database scaffolding
agent-frontmatter:end */

import { convertToSnakeCase } from "@agent-stack/utils";
import { type FieldAttribute, getTables } from "agent-stack/db";
import fs from "fs-extra";
import type { SchemaGenerator } from "./types";

type DrizzleProvider = "sqlite" | "pg" | "mysql";

export const generateDrizzleSchema: SchemaGenerator = async ({
  options,
  file,
  adapter,
}) => {
  const tables = getTables(options);
  const filePath = file ?? "./agent-schema.ts";

  const adapterOptions = (adapter.options ?? {}) as Partial<{
    provider: DrizzleProvider;
    usePlural: boolean;
  }>;
  const databaseType: DrizzleProvider = adapterOptions.provider ?? "sqlite";
  const usePlural = adapterOptions.usePlural ?? false;

  const importTokens: string[] = [];
  const useToken = (token: string) => {
    if (!importTokens.includes(token)) {
      importTokens.push(token);
    }
  };

  useToken(`${databaseType}Table`);

  const fileExist = fs.existsSync(filePath);

  // Map Agent Stack field metadata to drizzle column builders.
  const resolveField = (fieldName: string, field: FieldAttribute): string => {
    const dbName = convertToSnakeCase(fieldName);
    switch (field.type) {
      case "string": {
        if (databaseType === "mysql" && field.unique) {
          useToken("varchar");
          return `varchar('${dbName}', { length: 255 })`;
        }
        if (databaseType === "mysql" && field.references) {
          useToken("varchar");
          return `varchar('${dbName}', { length: 36 })`;
        }
        useToken("text");
        return `text('${dbName}')`;
      }
      case "boolean": {
        if (databaseType === "sqlite") {
          useToken("integer");
          return `integer('${dbName}', { mode: 'boolean' })`;
        }
        useToken("boolean");
        return `boolean('${dbName}')`;
      }
      case "number": {
        if (databaseType === "sqlite") {
          useToken("integer");
          return `integer('${dbName}')`;
        }
        if (field.bigint) {
          useToken("bigint");
          return `bigint('${dbName}', { mode: 'number' })`;
        }
        if (databaseType === "mysql") {
          useToken("int");
          return `int('${dbName}')`;
        }
        useToken("integer");
        return `integer('${dbName}')`;
      }
      case "date": {
        if (databaseType === "sqlite") {
          useToken("integer");
          return `integer('${dbName}', { mode: 'timestamp' })`;
        }
        useToken("timestamp");
        return `timestamp('${dbName}')`;
      }
      default: {
        if (field.type === "string[]") {
          useToken("text");
          return `text('${dbName}')`;
        }
        if (field.type === "number[]") {
          if (databaseType === "sqlite") {
            useToken("integer");
            return `integer('${dbName}')`;
          }
          if (databaseType === "mysql") {
            useToken("int");
            return `int('${dbName}')`;
          }
          useToken("integer");
          return `integer('${dbName}')`;
        }
        useToken("text");
        return `text('${dbName}')`;
      }
    }
  };

  const idLine = (() => {
    if (databaseType === "mysql") {
      useToken("varchar");
      return `varchar('id', { length: 36 }).primaryKey()`;
    }
    useToken("text");
    return `text('id').primaryKey()`;
  })();

  const tableBlocks: string[] = [];

  for (const [, table] of Object.entries(tables)) {
    const modelName = usePlural ? `${table.modelName}s` : table.modelName;
    const fieldLines = Object.entries(table.fields)
      .map(([fieldKey, attr]) => {
        if (fieldKey === "id") {
          return null;
        }
        const base = resolveField(fieldKey, attr);
        const notNull = attr.required ? ".notNull()" : "";
        const unique = attr.unique ? ".unique()" : "";
        const reference = attr.references
          ? `.references(() => ${
              usePlural ? `${attr.references.model}s` : attr.references.model
            }.${attr.references.field}, { onDelete: 'cascade' })`
          : "";
        return `  ${fieldKey}: ${base}${notNull}${unique}${reference},`;
      })
      .filter(Boolean) as string[];

    const schemaLines = [
      `export const ${modelName} = ${databaseType}Table("${convertToSnakeCase(
        modelName,
      )}", {`,
      `  id: ${idLine},`,
      ...fieldLines,
      `});`,
    ];

    tableBlocks.push(schemaLines.join("\n"));
  }

  const importLine = `import { ${importTokens.join(", ")} } from "drizzle-orm/${databaseType}-core";`;
  const code = `${importLine}\n\n${tableBlocks.join("\n\n")}\n`;

  return {
    code,
    fileName: filePath,
    overwrite: fileExist,
  };
};
