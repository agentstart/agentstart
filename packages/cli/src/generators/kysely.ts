import { getMigrations } from "@agent-stack/infra/db";
import type { SchemaGenerator } from "./types";

export const generateMigrations: SchemaGenerator = async ({
  options,
  file,
}) => {
  const { compileMigrations } = await getMigrations(options);
  const migrations = await compileMigrations();
  return {
    code: migrations,
    fileName:
      file ||
      `./@agent-stack/core_migrations/${new Date()
        .toISOString()
        .replace(/:/g, "-")}.sql`,
  };
};
