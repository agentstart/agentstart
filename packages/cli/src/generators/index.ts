import { logger } from "@agent-stack/utils";
import type { Adapter, AgentStackOptions } from "agent-stack";
import { generateDrizzleSchema } from "./drizzle";
import { generateMigrations } from "./kysely";
import { generatePrismaSchema } from "./prisma";

export const adapters = {
  prisma: generatePrismaSchema,
  drizzle: generateDrizzleSchema,
  kysely: generateMigrations,
};

export const getGenerator = (opts: {
  adapter: Adapter;
  file?: string;
  options: AgentStackOptions;
}) => {
  const adapter = opts.adapter;
  const generator =
    adapter.id in adapters
      ? adapters[adapter.id as keyof typeof adapters]
      : null;
  if (!generator) {
    logger.error(`${adapter.id} is not supported.`);
    process.exit(1);
  }
  return generator(opts);
};
