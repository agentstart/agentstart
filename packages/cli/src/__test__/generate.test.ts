/* agent-frontmatter:start
AGENT: CLI test module
PURPOSE: Exercises AgentStart CLI commands to prevent regressions.
USAGE: Executed with Vitest to validate CLI generators and configuration helpers.
EXPORTS: None
FEATURES:
  - Covers critical CLI workflows for schema generation
  - Uses snapshot assertions to track emitted files
SEARCHABLE: packages, cli, src, test, generate, vitest
agent-frontmatter:end */

import type { AgentStartOptions } from "agentstart";
import { drizzleAdapter, prismaAdapter } from "agentstart/db";
import Database from "better-sqlite3";
import { describe, expect, it } from "vitest";
import { generateDrizzleSchema } from "../generators/drizzle";
import { generateMigrations } from "../generators/kysely";
import { generatePrismaSchema } from "../generators/prisma";

describe("generate", () => {
  it("should generate prisma schema", async () => {
    const schema = await generatePrismaSchema({
      file: "test.prisma",
      adapter: prismaAdapter(
        {},
        {
          provider: "postgresql",
        },
      )({} as AgentStartOptions),
      options: {
        memory: prismaAdapter(
          {},
          {
            provider: "postgresql",
          },
        ),
      },
    });
    expect(schema.code).toMatchSnapshot();
  });

  it("should generate prisma schema for mongodb", async () => {
    const schema = await generatePrismaSchema({
      file: "test.prisma",
      adapter: prismaAdapter(
        {},
        {
          provider: "mongodb",
        },
      )({} as AgentStartOptions),
      options: {
        memory: prismaAdapter(
          {},
          {
            provider: "mongodb",
          },
        ),
      },
    });
    expect(schema.code).toMatchSnapshot();
  });

  it("should generate prisma schema for mysql", async () => {
    const schema = await generatePrismaSchema({
      file: "test.prisma",
      adapter: prismaAdapter(
        {},
        {
          provider: "mysql",
        },
      )({} as AgentStartOptions),
      options: {
        memory: prismaAdapter(
          {},
          {
            provider: "mongodb",
          },
        ),
      },
    });
    expect(schema.code).toMatchSnapshot();
  });

  it("should generate drizzle schema", async () => {
    const schema = await generateDrizzleSchema({
      file: "test.drizzle",
      adapter: drizzleAdapter(
        {},
        {
          provider: "pg",
          schema: {},
        },
      )({} as AgentStartOptions),
      options: {
        memory: drizzleAdapter(
          {},
          {
            provider: "pg",
            schema: {},
          },
        ),
      },
    });
    expect(schema.code).toMatchSnapshot();
  });

  it("should generate kysely schema", async () => {
    const schema = await generateMigrations({
      file: "test.sql",
      options: {
        memory: new Database(":memory:"),
      },
      adapter: {} as any,
    });
    expect(schema.code).toMatchSnapshot();
  });
});
