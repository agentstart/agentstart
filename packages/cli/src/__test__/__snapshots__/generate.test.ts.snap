// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate > should generate drizzle schema 1`] = `
"import { pgTable, text, timestamp } from "drizzle-orm/pg-core";

export const thread = pgTable("thread", {
  id: text('id').primaryKey(),
  title: text('title').notNull(),
  userId: text('user_id').notNull(),
  visibility: text('visibility').notNull(),
  lastContext: text('last_context'),
  createdAt: timestamp('created_at').notNull(),
  updatedAt: timestamp('updated_at').notNull(),
});

export const message = pgTable("message", {
  id: text('id').primaryKey(),
  threadId: text('thread_id').notNull().references(() => thread.id, { onDelete: 'cascade' }),
  role: text('role').notNull(),
  parts: text('parts').notNull(),
  attachments: text('attachments'),
  metadata: text('metadata'),
  createdAt: timestamp('created_at').notNull(),
  updatedAt: timestamp('updated_at').notNull(),
});

export const todo = pgTable("todo", {
  id: text('id').primaryKey(),
  threadId: text('thread_id').notNull().references(() => thread.id, { onDelete: 'cascade' }),
  todos: text('todos').notNull(),
  createdAt: timestamp('created_at').notNull(),
  updatedAt: timestamp('updated_at').notNull(),
});
"
`;

exports[`generate > should generate kysely schema 1`] = `
"create table "thread" ("id" text not null primary key, "title" text not null, "userId" text not null, "visibility" text not null, "lastContext" text, "createdAt" date not null, "updatedAt" date not null);

create table "message" ("id" text not null primary key, "threadId" text not null references "thread" ("id"), "role" text not null, "parts" text not null, "attachments" text, "metadata" text, "createdAt" date not null, "updatedAt" date not null);

create table "todo" ("id" text not null primary key, "threadId" text not null references "thread" ("id"), "todos" text not null, "createdAt" date not null, "updatedAt" date not null);"
`;

exports[`generate > should generate prisma schema 1`] = `
"
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Thread {
  id          String    @id
  title       String
  userId      String
  visibility  String
  lastContext String?
  createdAt   DateTime
  updatedAt   DateTime
  messages    Message[]
  todos       Todo[]

  @@map("thread")
}

model Message {
  id          String   @id
  threadId    String
  thread      Thread   @relation(fields: [threadId], references: [id], onDelete: Cascade)
  role        String
  parts       String
  attachments String?
  metadata    String?
  createdAt   DateTime
  updatedAt   DateTime

  @@map("message")
}

model Todo {
  id        String   @id
  threadId  String
  thread    Thread   @relation(fields: [threadId], references: [id], onDelete: Cascade)
  todos     String
  createdAt DateTime
  updatedAt DateTime

  @@map("todo")
}
"
`;

exports[`generate > should generate prisma schema for mongodb 1`] = `
"
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model Thread {
  id          String    @id @map("_id")
  title       String
  userId      String
  visibility  String
  lastContext String?
  createdAt   DateTime
  updatedAt   DateTime
  messages    Message[]
  todos       Todo[]

  @@map("thread")
}

model Message {
  id          String   @id @map("_id")
  threadId    String
  thread      Thread   @relation(fields: [threadId], references: [id], onDelete: Cascade)
  role        String
  parts       String
  attachments String?
  metadata    String?
  createdAt   DateTime
  updatedAt   DateTime

  @@map("message")
}

model Todo {
  id        String   @id @map("_id")
  threadId  String
  thread    Thread   @relation(fields: [threadId], references: [id], onDelete: Cascade)
  todos     String
  createdAt DateTime
  updatedAt DateTime

  @@map("todo")
}
"
`;

exports[`generate > should generate prisma schema for mysql 1`] = `
"
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model Thread {
  id          String    @id
  title       String    @db.Text
  userId      String    @db.Text
  visibility  String    @db.Text
  lastContext String?   @db.Text
  createdAt   DateTime
  updatedAt   DateTime
  messages    Message[]
  todos       Todo[]

  @@map("thread")
}

model Message {
  id          String   @id
  threadId    String
  thread      Thread   @relation(fields: [threadId], references: [id], onDelete: Cascade)
  role        String   @db.Text
  parts       String   @db.Text
  attachments String?  @db.Text
  metadata    String?  @db.Text
  createdAt   DateTime
  updatedAt   DateTime

  @@map("message")
}

model Todo {
  id        String   @id
  threadId  String
  thread    Thread   @relation(fields: [threadId], references: [id], onDelete: Cascade)
  todos     String
  createdAt DateTime
  updatedAt DateTime

  @@map("todo")
}
"
`;
